controller/root.go

```diff
@@ -0,0 +1,152 @@
+package controller
+
+import (
+	"encoding/json"
+	"strings"
+	"sync"
+
+	"github.com/fabric8-services/fabric8-wit/app"
+	"github.com/fabric8-services/fabric8-wit/errors"
+	"github.com/fabric8-services/fabric8-wit/jsonapi"
+	"github.com/fabric8-services/fabric8-wit/log"
+	"github.com/fabric8-services/fabric8-wit/ptr"
+	"github.com/fabric8-services/fabric8-wit/rest"
+	"github.com/fabric8-services/fabric8-wit/swagger"
+	"github.com/goadesign/goa"
+	errs "github.com/pkg/errors"
+	"github.com/satori/go.uuid"
+)
+
+const embeddedSwaggerSpecFile = "swagger.json"
+
+type asseter interface {
+	Asset(name string) ([]byte, error)
+}
+
+// RootController implements the root resource.
+type RootController struct {
+	*goa.Controller
+	FileHandler asseter
+	// endpoints caches the result of parsing the root endpoints so that
+	// consecutive calls to this API only call the API once
+	endpoints *app.Root
+	// endpointsLock protects the endpoints resource
+	endpointsLock sync.RWMutex
+}
+
+type workingFileFetcher struct{}
+
+func (s workingFileFetcher) Asset(fileName string) ([]byte, error) {
+	return swagger.Asset(fileName)
+}
+
+var _ asseter = workingFileFetcher{}
+var _ asseter = (*workingFileFetcher)(nil)
+
+// NewRootController creates a root controller.
+func NewRootController(service *goa.Service) *RootController {
+	return &RootController{
+		Controller:  service.NewController("RootController"),
+		FileHandler: workingFileFetcher{},
+	}
+}
+
+// List runs the list action.
+func (c *RootController) List(ctx *app.ListRootContext) error {
+	c.endpointsLock.Lock()
+	defer c.endpointsLock.Unlock()
+	if c.endpoints == nil {
+		roots, err := getRoot(ctx, c.FileHandler)
+		if err != nil || roots == nil {
+			log.Error(ctx, map[string]interface{}{
+				"err":  errs.WithStack(err),
+				"file": embeddedSwaggerSpecFile,
+			}, err.Error())
+			return jsonapi.JSONErrorResponse(ctx, errs.WithStack(err))
+		}
+		c.endpoints = roots
+	}
+	return ctx.OK(&app.RootSingle{Data: c.endpoints})
+}
+
+// Get a list of all endpoints formatted to json api format.
+func getRoot(ctx *app.ListRootContext, fileHandler asseter) (*app.Root, error) {
+	// Get an unmarshal swagger specification
+	swaggerJSON, err := fileHandler.Asset(embeddedSwaggerSpecFile)
+	if err != nil {
+		// TODO(tinakurian): fix error handling
+		return nil, errors.NewNotFoundError("file", embeddedSwaggerSpecFile)
+	}
+
+	var result map[string]interface{}
+	err = json.Unmarshal([]byte(swaggerJSON), &result)
+	if err != nil {
+		return nil, errs.Wrapf(err, "unable to unmarshal the file with id "+"'"+embeddedSwaggerSpecFile+"'")
+	}
+
+	// Get and iterate over paths from swagger specification.
+	swaggerPaths, ok := result["paths"]
+	if !ok {
+		return nil, errors.NewInternalErrorFromString("field `paths` could be found in swagger specification")
+	}
+
+	swaggerPathz, ok := swaggerPaths.(map[string]interface{})
+	if !ok {
```

Comment:
A *more* golang way of doing type assertions would be
```
      // Type assertion inside if
	var val interface{}
	val = "foo"
	if str, ok := val.(string); ok {
		fmt.Println(str)
	}
```

Reply:
No, this only works if `str` is not needed outside the if-block. And if the error handling is way shorter than the regular path, then it makes sense to put the error in the if like @tinakurian does here.

Reply:
@kwk Please see https://github.com/fabric8-services/fabric8-wit/pull/2169#discussion_r204683708. The variable `swaggerPathz` is not really needed. I agree this would not be useful if `str` is being used outside of the `if` but in this case, we do not need the `swaggerPathz`.

Reply:
I think I agree with Konrad, this would cause more nesting. Since we do need swaggerPathz to range over, it would need to be placed within the if condition.